### ISSUE TITLE
Set 5 and 6

### ISSUE THREAD

**Issue by Bishal-Kalita** [Bishal-Kalita/AssignmentDenvey#1]
Set 5
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
file_path = '/mnt/data/housing.csv'

try:
    data = pd.read_csv(file_path)
    print("Dataset loaded successfully!")
    print("First 5 rows:\n", data.head())
    print("\nData Info:\n")
    print(data.info())
    print("\nMissing values:\n", data.isnull().sum())
except Exception as e:
    print("Error loading dataset:", e)

data.dropna(inplace=True)

# Select features and target
# Assuming 'median_house_value' is the target variable
# and excluding any non-numeric or non-feature columns like 'ocean_proximity' if present
if 'median_house_value' in data.columns:
    X = data.select_dtypes(include=[np.number]).drop('median_house_value', axis=1)
    y = data['median_house_value']
else:
    raise ValueError("Target column 'median_house_value' not found in dataset.")

#  (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate model performance
r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print("\nModel Evaluation:")
print(f"R-squared score: {r2:.4f}")
print(f"Mean Squared Error: {mse:.2f}")

# Step 1: Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import seaborn as sns

# Step 2: Safe dataset loading with error handling
def load_dataset(filepath):
    try:
        data = pd.read_csv(filepath, sep=' ', header=None)
        data.columns = ['Feature1', 'Feature2']
        print("Dataset loaded successfully!")
        return data
    except FileNotFoundError:
        print("Error: File not found.")
        return None
    except pd.errors.ParserError:
        print("Error: Could not parse the file.")
        return None
    except Exception as e:
        print(f"Unexpected error: {e}")
        return None

# Load your dataset
file_path = '/content/a1.txt'  # Update if uploaded elsewhere
data = load_dataset(file_path)

# Step 3: Preprocessing
if data is not None:
    scaler = StandardScaler()
    scaled_data = scaler.fit_transform(data)

    # Step 4: Elbow method to determine optimal number of clusters
    distortions = []
    K = range(1, 11)
    for k in K:
        kmeans = KMeans(n_clusters=k, random_state=42)
        kmeans.fit(scaled_data)
        distortions.append(kmeans.inertia_)

    # Plot the Elbow graph
    plt.figure(figsize=(8, 5))
    plt.plot(K, distortions, 'bo-')
    plt.xlabel('Number of clusters')
    plt.ylabel('Inertia')
    plt.title('Elbow Method For Optimal k')
    plt.grid(True)
    plt.show()

    # Step 5: Choose optimal k (say k=3 based on elbow)
    optimal_k = 3
    kmeans = KMeans(n_clusters=optimal_k, random_state=42)
    clusters = kmeans.fit_predict(scaled_data)
    data['Cluster'] = clusters

    # Step 6: Visualization
    plt.figure(figsize=(8, 6))
    sns.scatterplot(x=data['Feature1'], y=data['Feature2'], hue=data['Cluster'], palette='Set1', s=60)
    plt.scatter(kmeans.cluster_centers_[:, 0]*scaler.scale_[0] + scaler.mean_[0],
                kmeans.cluster_centers_[:, 1]*scaler.scale_[1] + scaler.mean_[1],
                c='black', s=200, alpha=0.5, marker='X', label='Centroids')
    plt.title(f'K-Means Clustering (k={optimal_k})')
    plt.legend()
    plt.show()
